### ISSUE TITLE
Bug on calender calculation of EPI week

### ISSUE THREAD

**Issue by Iheanacho2027** [hzi-braunschweig/SORMAS-Project#3916]
### Bug Description
SORMAS calculates 2020 to have 52 weeks rather than 53 weeks known to be public. This has therefore affected the EPI week calender calculation for 2021 on SORMAS i.e EPI week 1 of 2021 which should begin from 4th January 2021 begins from 28th December 2020 on SORMAS 
### Steps to Reproduce
1.
2.

### System Details
* Device:
* SORMAS version:
* Android version/Browser:
* Server URL:
* User Role:

**Comment by MateStrysewske** [comment]
@Iheanacho2027 The specifications we got when we implemented the epi week system was specifically that the first epi week of a year is ALWAYS the week that contains the 1st of January. As far as I know, there are different epi week systems in use in the world, e.g. the system used in the US differs from the system used in some other countries. The system that SORMAS currently uses is what we were told to implement. @bernardsilenou If this is no longer valid, please forward a new definition to us; we'll have to check whether changing the calculations will lead to undesired side effects though.

**Comment by Marko-ilmkreis** [comment]
Usually the first week of the year is this week where the 1st January is on Monday, Tuesday, Wednesday or Thursday. (Source: https://en.wikipedia.org/wiki/Week#The_ISO_week_date_system)

**Comment by Iheanacho2027** [comment]
Thanks @Marko-ilmkreis
From the source you provided, "Week 1 of 2021 (2021W01 in ISO notation) will start on Monday, 4 January 2021 and will end on Sunday, 10 January 2021, because 1 January 2021 fell on Friday."

**Comment by MateStrysewske** [comment]
@Marko-ilmkreis Epi weeks aren't necessarily using the same system as ISO weeks though, that's why we used a different system.

**Comment by bernardsilenou** [comment]
@MateStrysewske I would agree with you and suggest that there is no need to change this in SORMAS now. What we have to keep consistent is to use the same definition each time  we do any computation by epi week.
@Iheanacho2027 If you need to do your reporting with the other type of epi week, then you can always export the data and use any grouping format.
Maybe when the thousands of issues would be over, we can add a parameter to configure the type of epi week to use in the system.

**Comment by bernardsilenou** [comment]
@Iheanacho2027 Please let us know if we can close the ticket in case there is no specific urgent request from your team?

**Comment by Iheanacho2027** [comment]
Pls do not close the ticket because it has begun to raise concerns in Nigeria. @Chinedar @foyiri

**Comment by bernardsilenou** [comment]
@Iheanacho2027 @Chinedar  @foyiri  They can also filter the dashboard by report date and choose the dates that fall in the week that they want to report. You can give them all alternative solution if someone complained. As we discussed yday, we should be working on having the the sample import module now rather changing date formats.

You guys are the once to decide on the final priority

**Comment by Chinedar** [comment]
@Silenou, Bernard <Bernard.Silenou@helmholtz-hzi.de> please we make both
priority. The calender affects the report they generate from the system.
Total reliance on SORMAS is what we are pushing for. This is critical to
it. Thanks.

**Comment by MateStrysewske** [comment]
@Iheanacho2027 @bernardsilenou @Chinedar So this means that we should use the official ISO week definition in the future, yes?

* We definitely need to keep in mind migration for weekly reports here.

**Comment by bernardsilenou** [comment]
@Chinedar I also agree we do the change. We would put it on the table for 99 and see if we can make it.
@MateStrysewske I would suggest that I have a look and compare the ISO week with other version (e.g the one i sent by email) before we can decide the one to use.
@Chinedar Since we are talking about week, then the day does not  matter but the day of the week matters. Please can you find out if Nigeria uses the version with epi-week that starts on a Monday or on a Sunday of each year?

I would suggest that when we decide on the system to use, then we have to also use the same thing for aggregate reporting. In this case, I would prefer the system that starts on a Monday to facilitate the match with weekly reporting

**Comment by bernardsilenou** [comment]
@MateStrysewske Is there something still needed from us? This ticket is classified as HZI urgent.

Here is the summary of my last email:

"The definition that the first epi week of a year should ALWAYS be the one that contains the 1st January, no matter on which day." is not the optimum definition is terms epidemiology.

The rule is that: The first day of week 1 of any year starts with either the first Sunday or Monday of the year.

1 and 2 January 2021 calendar year falls in week 52 of 2020.
Systems that always begins on a Sunday would put 3 Jan 2021 in Week 1 of 2020 while other that always start on Monday would put 3 Jan 2021 in week 52 of 2020.

Other factors to consider are: if the year is a leap year, has 365 or 366 days etc

In R documentations, we have packages that helps us to do this out of the box (https://rdrr.io/cran/epitools/man/as.week.html), I do not think we have to write a method to do this. 

In summary, we can use this package with the system that starts on a Monday to facilitate the match with weekly/zero reporting. We would not use this grouping only for epi-curve but for all default grouping by week.

**Comment by MateStrysewske** [comment]
@bernardsilenou So according to your definition, a year where the first Monday is on the 7th of January would also start week 1 at the 7th of January. That is definitely a huge difference compared to both the system that we used so far and also the ISO week system, which has an entirely different rule set.

Can you please send me a link to an official source that defines when the first epi week of a year starts and that is accepted by all countries that could potentially use SORMAS? We can't rely on vague definitions here, we need an official ruleset that we can rely on and that we can refer to.

**Comment by bernardsilenou** [comment]
@MateStrysewske  You are right. Such a year before this one would probably have 53 weeks instead of 52. Week 52 or 53 often cut across calendar days in both years.
Let me have a look from WHO and comments please.

**Comment by bernardsilenou** [comment]
@MateStrysewske I had a look around and found the same thing. The  CDC method `CDC (MMWR)` and WHO definition (see highlighted section is attached document) prefers the option where the week begins on a Sunday. So its better we use this method and adjust our zero reporting logic such that users would only get reminders on Mondays i.e second day of the new epi week.
We can use the method specified here (https://rdrr.io/cran/epitools/man/as.week.html) + first week  == Sunday

Refs, you can read just highlighted sections in both documents

[2016-cha-epidemiological-calendar.pdf](https://github.com/hzi-braunschweig/SORMAS-Project/files/5983973/2016-cha-epidemiological-calendar.pdf)
[MMWR_Week_overview.pdf](https://github.com/hzi-braunschweig/SORMAS-Project/files/5983998/MMWR_Week_overview.pdf)

**Comment by Marko-ilmkreis** [comment]
@bernardsilenou for the calculation of the first week of the year it is fine, but not for the normal calendar week. This week must start on Monday or Sunday, depending on the country. Otherwise it would not accepted by the users.

**Comment by MateStrysewske** [comment]
@Marko-ilmkreis Epi weeks always start at Monday or Sunday as well.
@bernardsilenou Wouldn't it be a better idea to allow system admins to define via properties whether to use Sunday or Monday as the first day of a week? I'm pretty sure that this is handled differently around the world, even for epi weeks.

**Comment by Marko-ilmkreis** [comment]
@MateStrysewske Thanks.

**Comment by bernardsilenou** [comment]
@MateStrysewske @Marko-ilmkreis  Perfect and better idea. Lets make the firstDayOfYear configurable: Sunday or Monday. We then use Sunday as default  configuration.

**Comment by MateStrysewske** [comment]
I recommend to create some weekly reports for the end of 2020 and the weeks that have passed in 2021 so far before starting development because we need to migrate at least those from 2021 and this needs to be tested.

- [ ] Add a new property to sormas.properties called `firstDayOfWeek`
    * Allowed values are "Sunday" and "Monday"
    * Default value is "Sunday"
    * Don't throw an error when an invalid value is entered, but fall back to the default instead
- [ ] Change methods in `DateHelper` that are using the first day of a week in some way. This includes (but doesn't necessarily extend to - please check for additional occurrences):
    * `getStartOfWeek` and `getEndOfWeek`
    * `getEpiCalendar`
    * `getEpiWeekStart` and `getEpiWeekEnd`
- [ ] Adjust code that calculates when the first epi week of a year starts according to the following definition: **The first epidemiological week of the year ends, by definition, on the first Saturday (or Sunday) of January, as long as it falls at least four days into the month, even if it means that this first week starts in December.** This includes (but doesn't necessarily extend to - please check for additional occurrences):
    * `DateHelper.getEpiCalendar`
    * the `epiWeek` and `epiYear` functions in `sormas_schema.sql` very likely need to be updated and recreated, and probably need an additional parameter that defines the first day of the week; if that's not possible, we need to have two functions for each, one for Monday and one for Sunday
- [ ] Check whether code that depends on this is still working as expected, e.g. `DateHelper.getEpiWeekWithCorrectYear`, the report forms in the web app and Android app, etc.
- [ ] Migrate weekly reports with year 2021 in the database:
    * Move reports for week 01-2021 to week 53-2020
    * Move all other reports for 2021 to one week earlier, e.g. 02-2021 to 01-2021, 03-2021 to 02-2021, and so on

@MartinWahnschaffe Anything else you can think of that I might have forgotten?

**Comment by MartinWahnschaffe** [comment]
I was a bit unsure whether we should enforce a fixed first day of the week for the whole server. On the other side I can't think of a use case where SORMAS would be used in multiple regions that have a different first day of the week, so it's probably ok.

For the property I'd suggest to use https://docs.oracle.com/javase/8/docs/api/java/time/DayOfWeek.html as the data type in the code and allow case-unsensitive casting of the string provided in the properties file.

Besides that I don't see anything obvious that is missing here.

**Comment by Iheanacho2027** [comment]
@bernardsilenou @MateStrysewske pls we agreed to make this configurable to system admins using a property called firstDayOfWeek. However this wasnt implemented as the first day of the week is set to default as monday in contradiction to what is applicable in Nigeria. Pls kindly help implement urgently

@Chinedar

**Comment by bernardsilenou** [comment]
@Iheanacho2027  @Chinedar 
This issue has not yet been implemented as you can see. Since much time has passed and things have also changed, It would be good that we have a meeting to discuss all  the urgent issues from you side and plan how to implement them.
You can suggest days in the coming weeks please.
Thanks

**Comment by bernardsilenou** [comment]
@markusmann-vg @MateStrysewske @MartinWahnschaffe 
- There may be a bug that is related to this ticket.
- The current computation of a week does not correctly compute the last or first week of a year, causing the whole series to be shifted. 
- When you filter the surveillance dashboard for cases reported between December 15, 2022, and January 31, 2023, on the test1401 or any other server, version 1.80, and group epi curve by week and subsequently by date, records are not assigned to the right week. 
- Some examples:
December 23, 2022 - Jan 1, 223 is computed as week 1 of 2023.  For Monday system, week 52 == 26/12/2022 -  01/01/2023 and for Sunday system, week 52 == 25/12/2022 - 31/12/2022

![image](https://user-images.githubusercontent.com/12996724/223565799-19ae140e-2a94-4173-877b-7768abe63355.png)

- This also reflects when grouping by date and by week of the same extracted data
- Group by date
![image](https://user-images.githubusercontent.com/12996724/223566956-a7f8b728-4e28-4b86-bd52-acac8cb07bab.png)
- group by month
- 
![image](https://user-images.githubusercontent.com/12996724/223567108-f4dfd04b-5881-4692-abc9-c99d4ffda6b2.png)

**Comment by MateStrysewske** [comment]
@bernardsilenou This is not a bug, but precisely the subject of this change request; when we implemented epi weeks in SORMAS, the definitions given to us were different than the definitions formulated in this change request. The week including the 1st of January being defined as epi week 1 is a result of the initial definition.
